sway/tree/container.c

```diff
@@ -132,42 +133,67 @@ struct sway_container *container_create(enum sway_container_type type) {
 	return c;
 }
 
-static void _container_destroy(struct sway_container *cont) {
-	if (cont == NULL) {
-		return;
-	}
-
-	wl_signal_emit(&cont->events.destroy, cont);
+static void container_workspace_free(struct sway_workspace *ws) {
+	list_foreach(ws->output_priority, free);
+	list_free(ws->output_priority);
+	ws->floating->destroying = true;
+	container_free(ws->floating);
+	free(ws);
+}
 
-	struct sway_container *parent = cont->parent;
-	if (cont->children != NULL && cont->children->length) {
-		// remove children until there are no more, container_destroy calls
-		// container_remove_child, which removes child from this container
-		while (cont->children != NULL && cont->children->length > 0) {
-			struct sway_container *child = cont->children->items[0];
-			ipc_event_window(child, "close");
-			container_remove_child(child);
-			_container_destroy(child);
-		}
-	}
-	if (cont->marks) {
-		list_foreach(cont->marks, free);
-		list_free(cont->marks);
-	}
-	if (parent) {
-		parent = container_remove_child(cont);
+void container_free(struct sway_container *cont) {
+	if (!sway_assert(cont->destroying,
+				"Tried to free container which wasn't marked as destroying")) {
+		return;
 	}
-	if (cont->name) {
-		free(cont->name);
+	if (!sway_assert(cont->instructions->length == 0,
+				"Tried to free container with pending instructions")) {
+		return;
 	}
-
+	free(cont->name);
 	wlr_texture_destroy(cont->title_focused);
 	wlr_texture_destroy(cont->title_focused_inactive);
 	wlr_texture_destroy(cont->title_unfocused);
 	wlr_texture_destroy(cont->title_urgent);
 
+	for (int i = 0; i < server.destroying_containers->length; ++i) {
```

Comment:
I didn't realize we allowed that syntax, but there are plenty around so I guess it's ok :)